bigquery/google/cloud/bigquery/dataset.py

```diff
@@ -283,28 +293,38 @@ def from_string(cls, dataset_id, default_project=None):
             >>> DatasetReference.from_string('my-project-id.some_dataset')
             DatasetReference('my-project-id', 'some_dataset')
 
+            >>> DatasetReference.from_string('prefix:my-project-id.some_dataset')
+            DatasetReference('my-project-id', 'some_dataset')
+
         Raises:
             ValueError:
                 If ``dataset_id`` is not a fully-qualified dataset ID in
                 standard SQL format.
         """
         output_dataset_id = dataset_id
         output_project_id = default_project
-        parts = dataset_id.split(".")
+        with_prefix = _W_PREFIX.match(dataset_id)
+        if with_prefix is None:
+            parts = dataset_id.split(".")
+        else:
+            parts = with_prefix.group("ref").split(".")
```

Comment:
I'm confused. What is this doing? I think the prefix needs to be part of the project ID, right?

Reply:
From the issue's `Stack trace`:
The error occurs due to the prefix `google.com:`. Previously the passed string was separated only by `.`, what led to `ValueError` raising because of the `len(parts) > 2` at `google.com:[project].ryan_dataset`. As I see here prefix is not the part of the `Project ID` itself. I was trying to find out, how could I parse the string and fulfill both the previous and new format. That is why I decided to use regular expressions. Now, with template's help, it will separate prefix and solve several situations:
* `string-project.string_dataset` - will pass the template successfully as it was before;
* `prefix:string-project.string_dataset` - will group the part without prefix and then will divide it;
* `string-project:string_dataset` - if the `default_project` was not defined raises `ValueError`;
* `google.com:project:dataset_id` - same as above.

> ValueError: Too many parts in dataset_id. Expected a fully-qualified dataset ID in standard SQL format. e.g. "project.dataset_id", got google.com:[project].ryan_dataset

Reply:
Right, but it appears to me that we're discarding the prefix? Is that correct?

Reply:
Yeah, that is what I was thinking before this conversation. So, now I'm a bit confused. I thought the prefix is an extra part and should be just removed. But if it is actually the part of the `Project ID`, I'll need to reconfigure the pattern.